File: python/tests/test_concentration.py
#!/usr/bin/env python3
"""
Script: test_concentration.py
Objective: Verify Python soil-mass, degradation, and concentration calculations.
Author: Yi Yu
Created: 2026-06-24
Last updated: 2026-06-24
Inputs: Fixed masses, fractions, depths, densities, and elapsed times.
Outputs: Pytest assertions.
Usage: python -m pytest python/tests/test_concentration.py -q
Dependencies: numpy, pytest, cltf
"""

import numpy as np

from cltf.concentration import (
    apply_elapsed_degradation,
    resident_concentration_ug_kg,
    soil_mass_kg_ha,
)


def test_concentration_arithmetic() -> None:
    soil_mass = soil_mass_kg_ha(0, 100, 1.3)
    assert soil_mass == 1.3e6
    assert resident_concentration_ug_kg(21.32, 1.0, soil_mass, 0.2) == 16.4


def test_effective_porosity_is_a_normalized_scale() -> None:
    baseline = resident_concentration_ug_kg(20, 0.5, 1e6, 0.2)
    doubled_porosity = resident_concentration_ug_kg(20, 0.5, 1e6, 0.4)
    assert doubled_porosity == baseline / 2


def test_elapsed_degradation_completes_mass_balance() -> None:
    probabilities = np.array([[0.4, 0.3, 0.3]])
    result = apply_elapsed_degradation(probabilities, np.array([100.0]), 0.01)
    np.testing.assert_allclose(result.sum(axis=1), 1.0, atol=1e-12)

File: R/tests/testthat/test-concentration.R
#!/usr/bin/env Rscript
# Script: test-concentration.R
# Objective: Verify degradation, soil-mass, and resident-concentration calculations.
# Author: Yi Yu
# Created: 2026-06-23
# Last updated: 2026-06-24
# Inputs: cltf mass and concentration functions.
# Outputs: Testthat assertions.
# Usage: Loaded by testthat::test_local("R", filter = "concentration").
# Dependencies: testthat, cltf

test_that("soil mass uses depth interval and bulk density", {
  expect_equal(
    soil_mass_kg_ha(0, 100, bulk_density_g_cm3 = 1.3),
    1.3e6
  )
  expect_equal(
    soil_mass_kg_ha(100, 300, bulk_density_g_cm3 = 1.3),
    2.6e6
  )
})

test_that("concentration conversion reproduces application-rate arithmetic", {
  result <- resident_concentration_ug_kg(
    application_rate_g_ha = 21.32,
    remaining_fraction    = 1,
    soil_mass_kg_ha       = 1.3e6,
    effective_porosity    = 0.2
  )

  expect_equal(result, 16.4, tolerance = 1e-10)
})

test_that("effective porosity is a normalized concentration scale", {
  base <- resident_concentration_ug_kg(20, 0.5, 1e6, 0.2)
  half <- resident_concentration_ug_kg(20, 0.5, 1e6, 0.4)

  expect_equal(half, base / 2)
})

test_that("degradation balance includes degraded mass", {
  result <- apply_elapsed_degradation(
    layer_probabilities = matrix(c(0.4, 0.3, 0.3), nrow = 1),
    time_days           = 100,
    decay_rate_day      = 0.01
  )

  expect_equal(rowSums(result), 1, tolerance = 1e-12)
})
